File: samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/infrastructure/ConversationNotificationPublisher.java
package systems.zlink.samples.supportchat.server.support.infrastructure;

import systems.zlink.samples.supportchat.server.support.actors.SupportActorDirectory;
import systems.zlink.samples.supportchat.server.support.actors.SupportUserActor;
import systems.zlink.samples.supportchat.server.support.application.AgentAssignmentService;
import systems.zlink.samples.supportchat.server.support.domain.Conversation;
import systems.zlink.samples.supportchat.shared.contracts.Messages;

import java.util.Map;

public final class ConversationNotificationPublisher {
    // --8<-- [start:doc-sc-roster-push]
    public void assigned(
            AgentAssignmentService.AvailableAgent agent,
            Conversation.Snapshot state,
            SupportActorDirectory directory) {
        directory
                .require(agent.rosterActorId())
                .push(
                        new Messages.ConversationAssignedNotify(
                                state.conversationId(), ConversationContracts.state(state)));
    }

    // --8<-- [end:doc-sc-roster-push]

    public void publish(
            Conversation.Change change,
            Map<String, SupportUserActor> participants,
            AgentAssignmentService assignment) {
        for (Conversation.Event event : change.events()) {
            Messages.ConversationState state = ConversationContracts.state(event.state());
            switch (event.kind()) {
                case ParticipantJoined -> {
                    publishExcept(
                            participants,
                            event.actorId(),
                            new Messages.ParticipantJoinedNotify(
                                    state.conversationId(),
                                    event.actorId(),
                                    ConversationContracts.role(event.role()),
                                    state));
                }
                // --8<-- [start:doc-sc-message-push]
                case MessageAppended ->
                        publishExcept(
                                participants,
                                event.actorId(),
                                new Messages.ChatMessageNotify(
                                        state.conversationId(),
                                        ConversationContracts.message(event.message()),
                                        state));
                // --8<-- [end:doc-sc-message-push]
                case TypingChanged ->
                        publishExcept(
                                participants,
                                event.actorId(),
                                new Messages.TypingChangedNotify(
                                        state.conversationId(),
                                        event.actorId(),
                                        Boolean.TRUE.equals(event.typing()),
                                        state));
                case Idle ->
                        publishAll(
                                participants,
                                new Messages.ConversationIdleNotify(state.conversationId(), state));
                case Closed -> {
                    publishAll(
                            participants,
                            new Messages.ConversationClosedNotify(state.conversationId(), state));
                    assignment.releaseConversation(state.conversationId());
                }
            }
        }
    }

    private static void publishExcept(
            Map<String, SupportUserActor> participants, String actorId, Object notification) {
        participants.forEach(
                (participantId, actor) -> {
                    if (!participantId.equals(actorId)) {
                        actor.push(notification);
                    }
                });
    }

    private static void publishAll(
            Map<String, SupportUserActor> participants, Object notification) {
        participants.values().forEach(actor -> actor.push(notification));
    }
}

File: samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/infrastructure/zlink/spots/conversationspot/notifications/ConversationNotificationPublisher.kt
package systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.spots.conversationspot.notifications

import systems.zlink.samples.kotlin.supportchat.server.support.domain.ConversationEvent
import systems.zlink.samples.kotlin.supportchat.server.support.domain.ConversationEventKind
import systems.zlink.samples.kotlin.supportchat.server.support.domain.ConversationSnapshot
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.ConversationContracts
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.actors.SupportUserActor
import systems.zlink.samples.kotlin.supportchat.shared.contracts.ChatMessageNotify
import systems.zlink.samples.kotlin.supportchat.shared.contracts.ConversationAssignedNotify
import systems.zlink.samples.kotlin.supportchat.shared.contracts.ConversationClosedNotify
import systems.zlink.samples.kotlin.supportchat.shared.contracts.ConversationIdleNotify
import systems.zlink.samples.kotlin.supportchat.shared.contracts.ConversationState
import systems.zlink.samples.kotlin.supportchat.shared.contracts.ParticipantJoinedNotify
import systems.zlink.samples.kotlin.supportchat.shared.contracts.TypingChangedNotify

class ConversationNotificationPublisher {
    suspend fun publish(events: List<ConversationEvent>, actors: Map<String, SupportUserActor>) {
        for (event in events) {
            publish(event, actors)
        }
    }

    // --8<-- [start:doc-sc-roster-push]
    suspend fun publishAssignedToRoster(roster: SupportUserActor, snapshot: ConversationSnapshot) {
        val state = ConversationContracts.toState(snapshot)
        roster
            .context()
            .boundSession()
            .send(ConversationAssignedNotify(state.conversationId, state))
            .submit()
    }

    // --8<-- [end:doc-sc-roster-push]

    private suspend fun publish(event: ConversationEvent, actors: Map<String, SupportUserActor>) {
        val state = ConversationContracts.toState(event.state)
        when (event.kind) {
            ConversationEventKind.ParticipantJoined ->
                publishParticipantJoined(event, state, actors)
            ConversationEventKind.MessageAppended -> publishMessage(event, state, actors)
            ConversationEventKind.TypingChanged -> publishTyping(event, state, actors)
            ConversationEventKind.Idle ->
                publishAll(actors.values) { actor ->
                    actor
                        .context()
                        .boundSession()
                        .send(ConversationIdleNotify(state.conversationId, state))
                        .submit()
                }
            ConversationEventKind.Closed ->
                publishAll(excludeActor(actors, event.actorId).values) { actor ->
                    actor
                        .context()
                        .boundSession()
                        .send(ConversationClosedNotify(state.conversationId, state))
                        .submit()
                }
        }
    }

    private suspend fun publishParticipantJoined(
        event: ConversationEvent,
        state: ConversationState,
        actors: Map<String, SupportUserActor>,
    ) {
        val actorId = event.actorId ?: error("Participant joined event requires actor id.")
        val role = event.role ?: error("Participant joined event requires role.")
        val customer = actors[state.customerActorId] ?: return
        if (customer.participantId == actorId) {
            return
        }
        customer
            .context()
            .boundSession()
            .send(
                ParticipantJoinedNotify(
                    state.conversationId,
                    actorId,
                    ConversationContracts.toRole(role),
                    state,
                )
            )
            .submit()
    }

    // --8<-- [start:doc-sc-message-push]
    private suspend fun publishMessage(
        event: ConversationEvent,
        state: ConversationState,
        actors: Map<String, SupportUserActor>,
    ) {
        val message = event.message ?: error("Message event requires a chat message.")
        val chatMessage = ConversationContracts.toMessage(message)
        publishAll(excludeActor(actors, message.senderActorId).values) { actor ->
            actor
                .context()
                .boundSession()
                .send(ChatMessageNotify(state.conversationId, chatMessage, state))
                .submit()
        }
    }

    // --8<-- [end:doc-sc-message-push]

    private suspend fun publishTyping(
        event: ConversationEvent,
        state: ConversationState,
        actors: Map<String, SupportUserActor>,
    ) {
        val actorId = event.actorId ?: error("Typing event requires actor id.")
        val isTyping = event.isTyping ?: error("Typing event requires typing state.")
        publishAll(excludeActor(actors, actorId).values) { actor ->
            actor
                .context()
                .boundSession()
                .send(TypingChangedNotify(state.conversationId, actorId, isTyping, state))
                .submit()
        }
    }

    private fun excludeActor(
        actors: Map<String, SupportUserActor>,
        actorId: String?,
    ): Map<String, SupportUserActor> =
        if (actorId == null) actors else actors.filterKeys { it != actorId }

    private suspend fun publishAll(
        actors: Collection<SupportUserActor>,
        publish: (SupportUserActor) -> Unit,
    ) {
        for (actor in actors) {
            publish(actor)
        }
    }
}
